[core] C++20 support

AndrewShkrob opened the pull request.

biodranik commented:
Is C++20 already supported on all our platforms and targets?

biodranik commented on transit/world_feed/world_feed.cpp:
```diff
@@ -1069,14 +1069,8 @@ std::pair<size_t, size_t> WorldFeed::ModifyShapes()
           if (!stopsOnLines.m_isValid)
           {
             m_lines.m_data.erase(lineId);
-            // todo: use std::erase_if after c++20
```
В коде ещё много таких мест, где можно использовать erase_if

AndrewShkrob commented:
> Is C++20 already supported on all our platforms and targets?

All checks passed, so yes

biodranik commented:
Does gcc support it well? From which version? We use it on Linux. And other contributors are actively using Linux. Will it require the installation of some custom newer gcc versions (and a lot of time/manual work)?

Ferenc- commented:
> Does gcc support it well? From which version? We use it on Linux. And other contributors are actively using Linux. Will it require the installation of some custom newer gcc versions (and a lot of time/manual work)?

Currently `gcc` is at version `13`, and [C++20 support is available since](https://gcc.gnu.org/projects/cxx-status.html#cxx20) version `8`. That version is also available on the prehistoric Ubuntu 18.04 LTS (the oldest distro know to be actively used by at least one user).  All it takes to install it is a simple `apt install gcc-8`.
And BTW, last time I have checked, organicmaps did not compile with gcc `7` anyway, `8.1` was the minimum,

vng commented on 3party/opening_hours/parse_months.cpp:
```diff
@@ -31,17 +31,17 @@ namespace osmoh
-       bind(&DateOffset::SetWDayOffsetPositive, _val, _a)]
+      [(bind(&DateOffset::SetWDayOffset, _val, _1),
+        bind(&DateOffset::SetOffset, _val, _2),
+        bind(&DateOffset::SetWDayOffsetPositive, _val, _a))]
```
What is the logic behind this? Why additional _tuple_ here?

AndrewShkrob replied:
It's not a tuple. Top-level comma expression in array subscript is deprecated. It means that we can't write
```cpp
arr[a, b, c]
```

In C++20 this expression is deprecated because in C++23 it is a multidimensional operator [] - https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2128r6.pdf
```cpp
// 2 x 3 array of arrays
array aa {array{1, 2, 3}, array{4, 5, 6}};
assert( (aa[0,1] == 2) );
assert( (aa[1,2] == 6) );
```

So, the fix for this warning is to cover it with `(...)`
```cpp
arr[(a, b, c)]
```

vng commented on indexer/feature_data.cpp:
```diff
@@ -341,7 +341,7 @@ bool FeatureParams::AddHouseNumber(string houseNumber)
 
   // Negative house numbers are not supported.
-  if (houseNumber.front() == '-' || houseNumber.find(u8"－") == 0)
+  if (houseNumber.front() == '-' || houseNumber.find(reinterpret_cast<char const *>(u8"－")) == 0)
```
I'd prefer to remove u8 at all here and below. reinterpret_cast is not a good solution.

vng commented on tracking/archive.hpp:
```diff
@@ -44,7 +44,10 @@ struct Restrictions
+  Limits(double minLat, double maxLat, double minLon, double maxLon)
+    : m_minLat(minLat), m_maxLat(maxLat), m_minLon(minLon), m_maxLon(maxLon)
+  {
+  }
```
Do we need additional ctor here?

AndrewShkrob replied:
Yes. Prior to C++20, the compiler would generate a deduction guide for a struct if it had no user-defined constructors, allowing the struct to be initialized using an initializer list. However, in C++20, a user-defined constructor is required to use aggregate initialization with an initializer list. This change could cause the code to fail to compile if the struct being initialized does not have a user-defined constructor.

We can leave the struct as
```cpp
struct Limits
{
 double m_minLat;
 double m_maxLat;
 double m_minLon;
 double m_maxLon;
};
```

And it won't require any additional ctor. But originally default ctor was deleted, so I decided to create an explicit ctor.

AndrewShkrob replied:
There are two possible solutions:
```cpp
// 1
struct Limits
{
 double m_minLat;
 double m_maxLat;
 double m_minLon;
 double m_maxLon;
};
// 2
struct Limits
{
 Limits(double minLat, double maxLat, double minLon, double maxLon);
 double m_minLat;
 double m_maxLat;
 double m_minLon;
 double m_maxLon;
};
```

I implemented the second one because originally default ctor was deleted.
But we can switch to the first solution

biodranik replied:
We try to check the old code and its usage if there is a need to touch it. The final change should ideally take into account the usage, not just blindly copy whatever properties the code had before. Because there were so many changes, and these properties may not be needed anymore.

vng commented:
- I suspect that we can make
```
enum path_commands_e : unsigned
enum cell_block_scale_e : unsigned
```
once, instead of making lots of conversions.

- Use regular quotes in include "std/boost_geometry.hpp" for our wrappers

AndrewShkrob commented:
>- I suspect that we can make in agg
>   ```cpp
>   enum path_commands_e : unsigned
>   enum cell_block_scale_e : unsigned
>   ```
>once, instead of making lots of conversions.

No. We have [-Wdeprecated-enum-enum-conversion] which means we're not able to compare different enums

AndrewShkrob commented:
Almost all warnings are fixed (even old ones)

* [Linux builds and tests (g++-12, gcc-12, Debug)](https://github.com/organicmaps/organicmaps/actions/runs/4930378591/jobs/8811218502#logs) - no warnings.
* [Linux builds and tests (g++-12, gcc-12, RelWithDebInfo)](https://github.com/organicmaps/organicmaps/actions/runs/4930378591/jobs/8811218601?pr=5093#logs) - some weird warnings left in 3party and std. Some warnings left in our code.
* [Linux no unity build](https://github.com/organicmaps/organicmaps/actions/runs/4930378591/jobs/8811218878#logs) - deprecations in protobuf. We should update it.
* [Linux builds and tests (clang++-14, clang-14, Debug)](https://github.com/organicmaps/organicmaps/actions/runs/4930378591/jobs/8811218696#logs) - deprecations in protobuf.
* [Linux builds and tests (clang++-14, clang-14, RelWithDebInfo)](https://github.com/organicmaps/organicmaps/actions/runs/4930378591/jobs/8811218787?pr=5093#logs) - deprecations in protobuf.
* [macOS builds and tests (Debug)](https://github.com/organicmaps/organicmaps/actions/runs/4930378592/jobs/8811218625#logs) - deprecations in icu. We should update it.
* [macOS builds and tests (RelWithDebInfo)](https://github.com/organicmaps/organicmaps/actions/runs/4930378592/jobs/8811218768#logs) - deprecations in icu.

I'll try to fix some more warnings in [Linux builds and tests (g++-12, gcc-12, RelWithDebInfo)](https://github.com/organicmaps/organicmaps/actions/runs/4930378591/jobs/8811218601?pr=5093#logs) but seems we have to update 3party libs.

I think that's it and we can merge this PR.

@biodranik @vng PTAL

biodranik commented:
Before merging C++20 into the master, we should test if iOS and Android builds are working properly on real devices.
As an alternative, we may merge some warning fixes that are C++17 compatible (if any) first.

It would be great to also update 3party libs separately to make sure that their warnings are properly handled too (or silenced if they're safe, or fixed upstream if necessary).

AndrewShkrob commented:
> As an alternative, we may merge some warning fixes that are C++17 compatible (if any) first.

All this warning fixes are compatible with C++17 (except for std::erase_if).

>Before merging C++20 into the master, we should test if iOS and Android builds are working properly on real devices.

What exactly you want to check? C++20 or my warning fixes? I checked Android and it works :)

vng commented:
- BTW, we still have -std=c++17 in OmimHelpers.cmake
- Check what setting is in ninja build for Android: ./android/nativeOutput/.../build.ninja

AndrewShkrob commented:
>BTW, we still have -std=c++17 in OmimHelpers.cmake

Done

>Check what setting is in ninja build for Android: ./android/nativeOutput/.../build.ninja

C++20

>I suspect that the current iOS xcodeproj structure is not based on root CMakeList.txt

I updated iOS project. Now it compiles with C++20. But weird error appeared:

<img width="475" alt="image" src="https://github.com/organicmaps/organicmaps/assets/10351358/114ff7e7-5f85-4e95-9408-8a2f9efa3e06">


Fixed with
```cpp
#include "std/target_os.hpp"
#ifdef OMIM_OS_IPHONE
#include <coroutine>
#endif

#include <set>
```

Do you have any idea how to solve it properly?

biodranik commented:
> Do you have any idea how to solve it properly?

It would be great to see the full error log from the compiler.

I suggest splitting the PR and merging everything that may run with C++17 until we test C++20 properly.

Dushistov commented:
> BTW, last time I have checked, organicmaps did not compile with gcc `7` anyway, `8.1` was the minimum,

I use gcc 7.5.0 to build organicmaps for my linux tablet. It uses custom linux distro based on debian,
and manufacture has no plans to update linux on it.

I built organicmaps time to time, last time I compiled dda8168861b3d6a628dd3c496f7f44da1263b5ec with some minor changes.
And changes required mainly because of my changes in fork, not because of some compilation problems.
So I would be really grateful if C++17 support continues.

P.S.

And by the way, C++17 is  still not fully supported by the way in organicmaps.
For example, `std::strto(|fd)` can be replaced by [std::from_chars](https://en.cppreference.com/w/cpp/utility/from_chars) that not depend on locale,
and  3party/fast_double_parser dependency can be dropped, because of  [std::from_chars](https://en.cppreference.com/w/cpp/utility/from_chars) .

biodranik commented:
@Dushistov thanks, is std::from_chars supported on iOS and Android 5.0 platforms properly?

And what does your fork do? :)

biodranik commented:
@Dushistov Looks like clang [does not support floating point conversion](https://www.reddit.com/r/cpp_questions/comments/zg6qlt/comment/izfcvse/) with std::from_chars.

AndrewShkrob commented:
I decided to leave C++17 right now. There are some errors with iOS build and still some warnings with 3party libs. I'd like to try to fix all warnings before moving to C++20.

So, no there are only warning fixes.

vng commented on drape_frontend/circles_pack_shape.cpp:
```diff
@@ -130,7 +130,7 @@ void CirclesPackHandle::SetPoint(size_t index, m2::PointD const & position, floa
 void CirclesPackHandle::Clear()
 {
-  memset(m_buffer.data(), 0, m_buffer.size() * sizeof(CirclesPackDynamicVertex));
+  m_buffer.clear();
```
Are you sure that this change is ok?

vng replied:
UPD: Not ok. resize called once in ctor.

vng commented on editor/opening_hours_ui.hpp:
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include "std/target_os.hpp"
```
Remove?

vng commented on generator/final_processor_country.cpp:
```diff
@@ -196,9 +196,9 @@ bool DoesBuildingConsistOfParts(FeatureBuilder const & fbBuilding,
-  static auto const buildingWithPartsClassifType = classificator.GetTypeByPath({"building", "has_parts"});
+  [[maybe_unused]] static auto const buildingClassifType = classificator.GetTypeByPath({"building"});
+  [[maybe_unused]] static auto const buildingPartClassifType = classificator.GetTypeByPath({"building:part"});
+  [[maybe_unused]] static auto const buildingWithPartsClassifType = classificator.GetTypeByPath({"building", "has_parts"});
```
Why?

vng commented on generator/final_processor_country.cpp:
```diff
@@ -295,7 +295,7 @@ void CountryFinalProcessor::AddFakeNodes()
 void CountryFinalProcessor::DropProhibitedSpeedCameras()
 {
-  static auto const speedCameraType = classif().GetTypeByPath({"highway", "speed_camera"});
+  [[maybe_unused]] static auto const speedCameraType = classif().GetTypeByPath({"highway", "speed_camera"});
```
I don't like this code garbage. What does compiler print here?

vng commented on qt/qt_common/map_widget.cpp:
```diff
@@ -504,7 +504,7 @@ void MapWidget::wheelEvent(QWheelEvent * e)
 
   QOpenGLWidget::wheelEvent(e);
-  m_framework.Scale(exp(e->delta() / 360.0), m2::PointD(L2D(e->x()), L2D(e->y())), false);
+  m_framework.Scale(exp(e->angleDelta().y() / 360.0), m2::PointD(e->position().x(), e->position().y()), false);
```
Please, keep it for Qt6 PR (you can check valid changes there). It's not obvious and the scrolling speed changes.

vng commented on std/target_os.hpp:
```diff
@@ -8,6 +8,8 @@
 #elif defined(__APPLE__)
   #include <TargetConditionals.h>
   #if (TARGET_OS_IPHONE > 0)
+    #include <coroutine>
```
Remove?

vng commented on traffic/traffic_info.cpp:
```diff
@@ -202,7 +202,7 @@ void TrafficInfo::CombineColorings(vector<TrafficInfo::RoadSegmentId> const & ke
   size_t numKnown = 0;
   size_t numUnknown = 0;
-  size_t numUnexpectedKeys = knownColors.size();
+  [[maybe_unused]] size_t numUnexpectedKeys = knownColors.size();
```
?

AndrewShkrob replied:
Used only in debug and gcc in release says that it's never used

biodranik replied:
It's better to use `#ifdef DEBUG` in such cases for clarity.

vng replied:
Yes, and write ASSERT_GREATER(numUnexpectedKeys--, 0, ());

vng commented on transit/world_feed/world_feed_tests/world_feed_tests.cpp:
```diff
@@ -311,25 +311,25 @@ UNIT_TEST(Transit_GTFS_ProjectStopToLine_MaxDistance)
+ *             +    |
+ *             |    +
+ *             +         *F
+ */
```
What is the reason here?

AndrewShkrob replied:
```cpp
//              C*  /       \ 


//       + /    \+/           \
```

problem in `\` symbols at the end of the line
multi-line comment [-Wcomment]

Dushistov commented:
> thanks, is std::from_chars supported on iOS and Android 5.0 platforms properly?
> 

I am not sure about organicmaps, but when I build and distribute my application (not related to organicmaps in any way),
then 
1. For Android, libc++ is normally included into `.apk`, so you can ship any version of libc++ regardless of the Android OS version.
2. For iOS in theory you can link with custom libc++ and not system one. But I never tried.

> And what does your fork do? :)

The reason for the fork is mainly to show additional information on the map from a separate database. Well, it also downloads maps from other servers and with the help of its own download manager, but the reason for this is integration with other components of my program that need to download large amounts of data not related to cartography.

And cartography is just one of the functions, not one of the most important one. That is, omim (and organicmaps) is used as an auxiliary library. The application has been in development for several years, but is still in the form of closed alpha, so I can’t really add anything more.

> Looks like clang does not support floating point conversion with std::from_chars.

This is a pity. Ages ago Microsoft offer to use it's MSVC STL code for libc++ implementation of `std::from_chars`: 

[link](https://reviews.llvm.org/D70631)
> Microsoft would be delighted if you used MSVC's from_chars (and you can just take it

And now 2023, and libc++ still has no c++17 support :(

I mainly use libstdc++ so I have no idea that libc++ in so bad shape.

vng commented on map/mwm_url.cpp:
```diff
@@ -209,7 +209,7 @@ void ParsedMapApi::ParseMapParam(std::string const & key, std::string const & va
     }
 
-    ApiPoint pt{.m_lat = lat, .m_lon = lon};
+    ApiPoint pt{.m_lat = lat, .m_lon = lon, .m_name = "", .m_id = "", .m_style = ""};
```
- What warning is here?
- Replace it with simple ```points.push_back({lat, lon})``` ?

AndrewShkrob commented:
@biodranik @vng PTAL. Resolved your comments.

vng commented on map/mwm_url.hpp:
```diff
@@ -21,11 +21,11 @@ namespace url_scheme
+  double m_lon = 0.0;
+  std::string m_name = "";
+  std::string m_id = "";
+  std::string m_style = "";
```
What warning is here? I can understand lat, lon, but std::string m_name = ""; looks strange.

biodranik replied:
This looks very strange. Default initialization should be enough. There should be another way to fix it (or the warning is false-positive).

AndrewShkrob replied:
When I remove default initializers from the struct.

```
In file included from /home/runner/work/organicmaps/organicmaps/build/map/CMakeFiles/map.dir/Unity/unity_0_cxx.cxx:37:
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp: In member function ‘void url_scheme::ParsedMapApi::ParseMapParam(const std::string&, const std::string&, std::vector<url_scheme::ApiPoint>&, bool&)’:
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp:212:21: warning: missing initializer for member ‘url_scheme::ApiPoint::m_name’ [-Wmissing-field-initializers]
  212 |     points.push_back({.m_lat = lat, .m_lon = lon});
      |     ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp:212:21: warning: missing initializer for member ‘url_scheme::ApiPoint::m_id’ [-Wmissing-field-initializers]
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp:212:21: warning: missing initializer for member ‘url_scheme::ApiPoint::m_style’ [-Wmissing-field-initializers]
```

I think we can ignore this warning right now.

biodranik replied:
Can you locally silence it?

vng replied:
Is this warning also happens when your write points.push_back({lat, lon}); without var specifiers?

AndrewShkrob replied:
Still exists:
```
In file included from /home/runner/work/organicmaps/organicmaps/build/map/CMakeFiles/map.dir/Unity/unity_0_cxx.cxx:37:
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp: In member function ‘void url_scheme::ParsedMapApi::ParseMapParam(const std::string&, const std::string&, std::vector<url_scheme::ApiPoint>&, bool&)’:
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp:212:21: warning: missing initializer for member ‘url_scheme::ApiPoint::m_name’ [-Wmissing-field-initializers]
  212 |     points.push_back({lat, lon});
      |     ~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp:212:21: warning: missing initializer for member ‘url_scheme::ApiPoint::m_id’ [-Wmissing-field-initializers]
/home/runner/work/organicmaps/organicmaps/map/mwm_url.cpp:212:21: warning: missing initializer for member ‘url_scheme::ApiPoint::m_style’ [-Wmissing-field-initializers]
```

biodranik replied:
This is a well-known controversial warning. Let's disable it here locally for gcc-only using ifdef and pragma.

vng replied:
Hm, we should have a lot of warnings like this in this case. I'd prefer to not bother about it and don't put any pragmas garbage in the code.

biodranik replied:
Non-silenced warnings will hide some important warnings. Explicit gcc silencing will help to make it clearer in the future. There are not so many similar warnings in the code, right?

AndrewShkrob replied:
> Hm, we should have a lot of warnings like this in this case

>There are not so many similar warnings in the code, right?

It appears only here.

>I'd prefer to not bother about it and don't put any pragmas garbage in the code.

I agree. I don't like these pragmas.

> Non-silenced warnings will hide some important warnings

There are only 2-3 warnings left. Nothing will be hidden. I'd prefer to leave this warning for now and come back to it later. Maybe during the migration to C++20

vng reviewed (approved):
Please, add new changes in separate commits, will make squash then.
I've tired to read the whole PR after each change.
